shared/actions/chat2/index.js

```diff
@@ -860,43 +860,38 @@ const loadMoreMessages = (
     return actions
   }
 
-  const makeCall = RPCChatTypes.localGetThreadNonblockRpcSaga({
-    incomingCallMap: {
-      'chat.1.chatUi.chatThreadCached': p => onGotThread(p, 'cached'),
-      'chat.1.chatUi.chatThreadFull': p => onGotThread(p, 'full'),
-    },
-    params: {
-      cbMode: RPCChatTypes.localGetThreadNonblockCbMode.incremental,
-      conversationID,
-      identifyBehavior: RPCTypes.tlfKeysTLFIdentifyBehavior.chatGui,
-      pagination: {
-        last: false,
-        next: isScrollingBack ? 'deadbeef' : '', // daemon treats this as a boolean essentially. string means to scroll back, null means an initial load
-        num: numberOfMessagesToLoad,
-        previous: '',
-      },
-      pgmode: RPCChatTypes.localGetThreadNonblockPgMode.server,
-      query: {
-        disableResolveSupersedes: false,
-        markAsRead: false,
-        messageTypes: loadThreadMessageTypes,
-      },
-      reason: reasonToRPCReason(reason),
-    },
-    waitingKey: loadingKey,
+  return Saga.call(function*() {
+    try {
+      const results: RPCChatTypes.NonblockFetchRes = yield RPCChatTypes.localGetThreadNonblockRpcSaga({
+        incomingCallMap: {
+          'chat.1.chatUi.chatThreadCached': p => onGotThread(p, 'cached'),
+          'chat.1.chatUi.chatThreadFull': p => onGotThread(p, 'full'),
+        },
+        params: {
+          cbMode: RPCChatTypes.localGetThreadNonblockCbMode.incremental,
+          conversationID,
+          identifyBehavior: RPCTypes.tlfKeysTLFIdentifyBehavior.chatGui,
+          pagination: {
+            last: false,
+            next: isScrollingBack ? 'deadbeef' : '', // daemon treats this as a boolean essentially. string means to scroll back, null means an initial load
+            num: numberOfMessagesToLoad,
+            previous: '',
+          },
+          pgmode: RPCChatTypes.localGetThreadNonblockPgMode.server,
+          query: {
+            disableResolveSupersedes: false,
+            markAsRead: false,
+            messageTypes: loadThreadMessageTypes,
+          },
+          reason: reasonToRPCReason(reason),
+        },
+        waitingKey: loadingKey,
+      })
+      yield Saga.put(Chat2Gen.createSetConversationOffline({conversationIDKey, offline: results.offline}))
+    } finally {
+      yield Saga.put(Chat2Gen.createClearLoading({key: `pushLoad:${conversationIDKey}`}))
+    }
   })
-  return Saga.all([
-    Saga.identity(conversationIDKey),
-    makeCall,
-    Saga.put(Chat2Gen.createClearLoading({key: `pushLoad:${conversationIDKey}`})),
-  ])
-}
-
-const loadMoreMessagesSuccess = (results: ?Array<any>) => {
-  if (!results) return
-  const conversationIDKey: Types.ConversationIDKey = results[0]
-  const res: RPCChatTypes.NonblockFetchRes = results[1]
-  return Saga.put(Chat2Gen.createSetConversationOffline({conversationIDKey, offline: res.offline}))
```

Comment:
the actual bug is res.offline could be called on undefined res. With the new helper this never happens as the creation of results and the handling of it happen in one place instead of across 2 different functions

Reply:
Could you help me understand how `loadMoreMessageSuccess` was called when the PRC failed? Saga.all fails if any of it fails, so this shouldn't have been called if the RPC fails right?

Reply:
the return value of the rpc is not the only way it goes into the success callback. this is a large reason for the new saga helpers. at the top of the file i check values and bail if something isn't correct. in those cases i'm just returning, not throwing so it will go into the success callback but with no results. this is confounded even further if the rpc itself returns nothing on success (some do) so you really cant differentiate. this is why the pure pattern of 
```
takeEveryPure(patter, p, success, fail)
```
i think isn't great as it really disconnects the context of p and what happens after. also flow doesn't help you at all as they're basically disconnected things.

instead the new helper says
```
safeTakeEveryPurePromise(pattern, functionThatReturnsAPromiseToAnAction)
```
so its very clear what happens next

Reply:
👍🏿thanks for explaining!